pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,323 @@
+package virtdhcp
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"text/template"
+
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const (
+	DNSmasqExec = "/usr/sbin/dnsmasq"
+	DHCPPidFile = "/tmp/dhcp.pid"
+)
+
+var DHCPLeaseFile = "/var/run/kubevirt/dhcp.leases"
+var DHCPHostsFile = "/var/run/kubevirt/dhcp.hosts"
+
+type hostsDetails struct {
+	Mac   string
+	Lease int
+}
+
+type DNSmasqInstance struct {
+	hosts    map[string]hostsDetails
+	ipRange  [][]byte
+	confpath string
+	args     []string
+	cmd      *exec.Cmd
+	monitor  Monitor
+}
+
+func NewDNSmasq() *DNSmasqInstance {
+	dnsMasq := new(DNSmasqInstance)
+	dnsMasq.hosts = make(map[string]hostsDetails)
+	dnsMasq.ipRange = make([][]byte, 2)
+	dnsMasq.args = []string{"-k", "-d", "--strict-order", "--bind-dynamic", "--no-resolv", "--dhcp-no-override", "--dhcp-authoritative",
+		"--except-interface=lo", "--dhcp-hostsfile=" + DHCPHostsFile, "--dhcp-leasefile=" + DHCPLeaseFile,
+		"--pid-file=" + DHCPPidFile}
+	dnsMasq.monitor = NewMonitor()
+	OS = &OSHandler{}
+	return dnsMasq
+}
+
+// implement `Interface` in sort package.
+type sortByteArrays [][]byte
+
+func (b sortByteArrays) Len() int {
+	return len(b)
+}
+
+func (b sortByteArrays) Less(i, j int) bool {
+	// bytes package already implements Comparable for []byte.
+	switch bytes.Compare(b[i], b[j]) {
+	case -1:
+		return true
+	case 0, 1:
+		return false
+	default:
+		return false
+	}
+}
+
+func (b sortByteArrays) Swap(i, j int) {
+	b[j], b[i] = b[i], b[j]
+}
+
+type OSOps interface {
+	IsFileExist(path string) (bool, error)
+	getPidFromFile(file string) (int, error)
+	checkProcessExist(pid int) (bool, *os.Process)
+	killProcessIfExist(pid int) error
+	RecreateHostsFile() (*os.File, error)
+	executeCommand(string, []string) *exec.Cmd
+	readFromFile(file string) ([]byte, error)
+	IsDNSMasqPidRunning(pid int) bool
+}
+
+type OSHandler struct{}
+
+var OS OSOps
+
+func (o *OSHandler) IsFileExist(path string) (bool, error) {
+	_, err := os.Stat(path)
+	exists := false
+
+	if err == nil {
+		exists = true
+	} else if os.IsNotExist(err) {
+		err = nil
+	}
+	return exists, err
+}
+
+func (o *OSHandler) readFromFile(file string) ([]byte, error) {
+	content, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to open file: %s", file)
+		return nil, err
+	}
+	return content, nil
+}
+
+func (o *OSHandler) getPidFromFile(file string) (int, error) {
+	content, err := OS.readFromFile(file)
+	if err != nil {
+		return -1, err
+	}
+	lines := strings.Split(string(content), "\n")
+	pid, _ := strconv.Atoi(lines[0])
+	return pid, nil
+}
+
+func (o *OSHandler) killProcessIfExist(pid int) error {
+	procExist, proc := OS.checkProcessExist(pid)
+	if procExist {
+		killErr := proc.Kill()
```

Comment:
If you ensure that the dhcp server is always a child of the main process, you can just do a dump `proc.Kill()`, followed by a `proc.Exited()`, if an error occurred, to see if we saw the error because the process is already gone.

Reply:
dhcp is guaranteed to always be a child process now. This logic just exists to kill a pre-existing process (which will never actually happen with how virt-dhcp is being operated in a container at the moment)